File: src/analyzer/ComplexityAnalyzer.Core/EvidencePruner.cs
namespace ComplexityAnalyzer.Core;

/// <summary>
/// Drops O(1) placeholders that do not affect the reported bound.
/// Empty statement lists, locals, and literals become <c>empty: 1</c>
/// during composition; they are absorbed algebraically and should not
/// appear in the derivation tree.
/// </summary>
public static class EvidencePruner
{
    public static ComplexityEvidence Prune(ComplexityEvidence evidence)
    {
        var children = evidence.Children
            .Select(Prune)
            .Where(c => !IsNoise(c))
            .ToArray();

        if (evidence.Kind == "sequence" && children.Length == 1)
            return children[0];

        return evidence with { Children = children };
    }

    public static ComplexityEvidence Sequence(
        ComplexityExpression time,
        LineSpan? span,
        IEnumerable<ComplexityEvidence> parts)
    {
        var children = parts
            .Select(Prune)
            .Where(c => !IsNoise(c))
            .ToArray();
        if (children.Length == 1)
            return children[0];
        if (children.Length == 0)
        {
            return ComplexityEvidence.Leaf(
                "sequence", "empty", time, span);
        }

        return new ComplexityEvidence(
            "sequence",
            "sequential statements",
            time,
            span,
            children);
    }

    public static IReadOnlyList<ComplexityEvidence> Meaningful(
        IEnumerable<ComplexityEvidence> parts) =>
        parts.Select(Prune).Where(c => !IsNoise(c)).ToArray();

    public static bool IsNoise(ComplexityEvidence evidence)
    {
        if (evidence.Label == "empty") return true;
        return evidence.Kind == "sequence"
            && evidence.Children.Count == 0
            && evidence.Cost is ConstantExpression { Value: 1 };
    }
}

File: src/extension/src/analysis/typescript/engine/evidencePruner.ts
import type { ComplexityExpression } from './expression';
import { evidenceLeaf, type ComplexityEvidence, type LineSpan } from './types';

export function prune(evidence: ComplexityEvidence): ComplexityEvidence {
  const children = evidence.children
    .map(prune)
    .filter((child) => !isNoise(child));
  if (evidence.kind === 'sequence' && children.length === 1) {
    return children[0];
  }
  return { ...evidence, children };
}

export function sequenceEvidence(
  time: ComplexityExpression,
  span: LineSpan | undefined,
  parts: Iterable<ComplexityEvidence>,
): ComplexityEvidence {
  const children = [...parts]
    .map(prune)
    .filter((child) => !isNoise(child));
  if (children.length === 1) return children[0];
  if (children.length === 0) {
    return evidenceLeaf('sequence', 'empty', time, span);
  }
  return {
    kind: 'sequence',
    label: 'sequential statements',
    cost: time,
    span,
    children,
  };
}

export function meaningful(
  parts: Iterable<ComplexityEvidence>,
): ComplexityEvidence[] {
  return [...parts].map(prune).filter((child) => !isNoise(child));
}

export function isNoise(evidence: ComplexityEvidence): boolean {
  if (evidence.label === 'empty') return true;
  return evidence.kind === 'sequence'
    && evidence.children.length === 0
    && evidence.cost.kind === 'const'
    && evidence.cost.value === 1;
}
